### PR TITLE
NoImplicitAny: Transaction.ts

### PR DESCRIPTION
**Pull request by JCIV-SE**
Fixes #<issue_number_goes_here> (it's a good idea to open an issue first for discussion)

- [ ] Tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
     "datastore"
   ],
   "scripts": {
+    "all-test": "npm run test && npm run system-test",
```
Comment (praveenqlogic) [inline package.json]: avoid this while creating PR with google!

File: src/transaction.ts [hunk 2]
```diff
@@ -44,10 +48,10 @@ class Transaction extends DatastoreRequest {
   projectId: string;
   namespace?: string;
   readOnly: boolean;
-  request;
-  modifiedEntities_;
+  request: Request;
+  modifiedEntities_: ModifiedEntities_;
```
Comment (praveenqlogic) [inline src/transaction.ts]: type look better without underscore **ModifiedEntities**, what you say?

File: src/transaction.ts [hunk 3]
```diff
@@ -126,13 +130,24 @@ class Transaction extends DatastoreRequest {
...
+      gaxOptionsOrCallback = {};
+    } 
+    else if (typeof gaxOptionsOrCallback === 'object' && gaxOptionsOrCallback !== null) {} //!
+    else gaxOptionsOrCallback = {};
+
+    const gaxOptions: CommitGaxOptions = gaxOptionsOrCallback;
+    let callback: CommitCallback;
+
+    if (is.fn(cb) && cb !== undefined) callback = cb;
+    else callback = (()=>{});
```
Comment (praveenqlogic) [inline src/transaction.ts]: commit(gaxOptions?: CommitGaxOptions): Promise&lt;CommitResponse&gt;
 commit(callback: CommitCallback): void; 
commit(gaxOptions: CommitGaxOptions, callback: CommitCallback): void; 
commit(gaxOptionsOrCallback?: CommitGaxOptions|CommitCallback, callback?: CommitCallback): void|Promise&lt;CommitResponse&gt; 

**you can use below, line of code to do the things in two line**
const gaxOptions=
        typeof gaxOptionsOrCallback === 'object' ? gaxOptionsOrCallback : {};
    callback =
        typeof gaxOptionsOrCallback === 'function' ? gaxOptionsOrCallback : callback;

File: src/transaction.ts [hunk 4]
```diff
@@ -178,7 +191,7 @@ class Transaction extends DatastoreRequest {
         // callback, having all the keys together is necessary to maintain
         // order.
         .reduce(
-            (acc, entityObject) => {
+            (acc: Entity, entityObject) => {
```
Comment (praveenqlogic) [inline src/transaction.ts]: its look like acc is Array/collection, please verify it once!

File: src/transaction.ts [hunk 5]
```diff
@@ -278,7 +291,9 @@ class Transaction extends DatastoreRequest {
    *   });
    * });
    */
-  createQuery(namespace: string, kind?: string) {
+  createQuery(namespace: string): Query;
+  createQuery(namespace: string, kind: string): Query;
+  createQuery(namespace: string, kind?: string): Query {
```
Comment (praveenqlogic) [inline src/transaction.ts]: i think we can skip function overloading here due to optional parameter, any how both serving the same purpose!

File: src/transaction.ts [hunk 6]
```diff
@@ -314,7 +329,8 @@ class Transaction extends DatastoreRequest {
    *   });
    * });
    */
-  delete(entities) {
+  delete(entities: Entities): void;
+  delete(entities: Entities): void {
```
Comment (praveenqlogic) [inline src/transaction.ts]: no  need of overloading here!

File: src/transaction.ts [hunk 7]
```diff
@@ -360,19 +376,27 @@ class Transaction extends DatastoreRequest {
...
    * });
    */
-  rollback(gaxOptions, callback?) {
-    if (is.fn(gaxOptions)) {
-      callback = gaxOptions;
-      gaxOptions = {};
+  rollback(gaxOptions: RollbackGaxOptions): Promise<RollbackResponse>;
+  rollback(callback: RollbackCallback): void;
+  rollback(gaxOptionsOrCallback: RollbackGaxOptions|RollbackCallback, cb?: RollbackCallback): void|Promise<RollbackResponse> {
+    if (is.fn(gaxOptionsOrCallback) && typeof gaxOptionsOrCallback !== 'object') {
```
Comment (praveenqlogic) [inline src/transaction.ts]: nicely done 👍 , only add one more overload function
rollback(gaxOptions: RollbackGaxOptions, callback: RollbackCallback):void;

File: src/transaction.ts [hunk 8]
```diff
@@ -430,14 +454,20 @@ class Transaction extends DatastoreRequest {
...
    * });
    */
-  run(options?, callback?) {
-    if (is.fn(options)) {
-      callback = options;
-      options = {};
+  run(options?: RunOptions): Promise<RunCallback>;
+  run(callback?: RunCallback): void;
+  run(options?: RunOptions, callback?: RunCallback): void;
+  run(optionsOrCallback?: RunOptions|RunCallback|Entity, cb?: RunCallback): void|Promise<RunCallback> {
```
Comment (praveenqlogic) [inline src/transaction.ts]: in second and third overload parameters should not optional!

File: src/transaction.ts [hunk 9]
```diff
@@ -604,7 +634,8 @@ class Transaction extends DatastoreRequest {
    *   });
    * });
    */
-  save(entities) {
+  save(entities: Entities): void;
+  save(entities: Entities): void {
```
Comment (praveenqlogic) [inline src/transaction.ts]: overloading not needed!

File: package.json (hunk 1 above)
Comment (JCIV-SE) [reply package.json]: okay!

File: src/transaction.ts (hunk 2 above)
Comment (JCIV-SE) [reply src/transaction.ts]: It does, changed!

**Pull request merged**